### ISSUE TITLE
Error when the same class name is defined in except/else

### ISSUE THREAD

**Issue by lukaslihotzki-f** [microsoft/pyright#10170]
Functions with the same name can be defined both in if/else blocks and except/else blocks without errors. Classes with the same name can be defined in if/else blocks without errors. However, pyright errors when they are defined in except/else blocks, even if both kind of blocks are mutually exclusive.

https://github.com/microsoft/pyright/issues/2136 seems loosely related.

**Code or Screenshots**

The issue is specifically in class Foo, the others are just for comparison. In the whole code, pyright only reports a single error.

```python
try:
    pass
except:
    class Foo:
        pass
else:
    class Foo:
        pass

if True:
    class Bar:
        pass
else:
    class Bar:
        pass

try:
    pass
except:
    def baz():
        pass
else:
    def baz():
        pass
```

```
error: Class declaration "Foo" is obscured by a declaration of the same name (reportRedeclaration)
```

I reproduced this issue in the command line tool, using pyright 1.1.397.

**Comment by erictraut** [comment]
Pyright is working as designed here, so this isn't a bug. You're redeclaring a symbol twice with two incompatible types. This is no different than declaring two classes named `Foo` right after each other in the same block. It's also analogous to declared a variable `x: int` and then redeclaring it later in the same scope as `x: str`. Redeclarations like this are not allowed. A symbol can have only one declared type.

The reason you can declare `baz` twice is because the type declared by both `def` statements is the same.

For more details about type declarations, see [this documentation](https://microsoft.github.io/pyright/#/type-concepts?id=type-declarations).

If you'd like this to type check without error, you could declare two separate classes and then assign them conditionally to a variable, which would then have a union type.

```python
try:
    pass
except:
    class _Foo1:
        pass
    Foo = _Foo1
else:
    class _Foo2:
        pass
    Foo = _Foo2

reveal_type(Foo) # type[_Foo1] | type[_Foo2]
```
